feat(plugin): load dev plugin from memory

kurikomoe opened the pull request:
这个 feature 先存疑叭。

设计上是将所有的 dev plugin 加载到内存中，防止占用文件（重启后自动删除）。
这样子可以避免重启应用，热重载开发模式的 plugin。

可能的问题：
1. dev mode 加载的插件存储的数据不会自动删除（可以添加一个 database cleanup task，启动的时候自动扫描所有无关联 plugin 的数据并删除）
2. 如果 plugin 启动了持久的 bgtask，这些 task 不会被 cancel（这一点应该在开发者手册里面写明，有持久性任务的 plugin 必须冷重启应用之类的）

GoldenPotato137 commented:
就算崩溃后这些数据也可以被第二次加载进来的时候删除掉呀，我觉得应该不会造成特别大的问题。

热重载按钮确实可以有，或许可以考虑在删除的上面新增一行只在dev插件中可见的配置，里面一个reload按钮

kurikomoe committed:
feat(plugin): enhance dev mode support and UI display

- feat: load dev plugins from memory to support hot reload
- feat: sort plugins in UI (Dev mode > Name)
- feat: highlight dev plugins with red border
- fix: resolve plugin duplicate loading issue

kurikomoe committed:
add: plugin assembly unload

GoldenPotato137 commented on GalgameManager/Services/PluginService/PluginService.cs:
```diff
@@ -15,6 +15,21 @@
+        return string.Compare(x.Info.Name, y.Info.Name, StringComparison.OrdinalIgnoreCase);
+    }
+}
+
```
话说我们把比较函数丢在PluginX这个文件里会不会看起来逻辑更清晰一些，c++的friend函数都是丢在类里的，我们丢到同一个文件里看起来make sense一点（

GoldenPotato137 commented on GalgameManager/Services/PluginService/PluginService.cs:
```diff
@@ -23,32 +38,141 @@ public partial class PluginService(
+            i++;
+        }
+        _plugins.Insert(i, plugin);
+    }
```
我记得ObservableCollection好像支持设置比较函数来着，应该是不需要显式写一个插入函数的

kurikomoe replied:
应该是不支持的，我直接在 view 层做排序了。

GoldenPotato137 commented on GalgameManager/Services/PluginService/PluginService.cs:
```diff
@@ -23,32 +38,141 @@ public partial class PluginService(
+    {
+        _plugins.Remove(plugin);
+    }
+    
```
我觉得可预见的未来都不应该有其他人去直接从列表remove插件，我觉得还是直接内联在函数里比较好，毕竟直接从列表remove插件属于dev插件的特有需求，这么直接写个函数出来容易让其他开发者误调用它

GoldenPotato137 commented on GalgameManager/Contracts/Services/IPluginService.cs:
```diff
@@ -10,13 +10,27 @@ public interface IPluginService
+    /// </summary>
+    /// <param name="plugin"></param>
+    public void PluginDeleteData(PluginX plugin); 
+    
```
按字典序来放接口吧，（虽然别的Iterface因为石山并不遵守这个规则，我们新文件还是遵守规则好看一点吧

GoldenPotato137 commented on GalgameManager/Services/PluginService/PluginService.cs:
```diff
@@ -23,32 +38,141 @@ public partial class PluginService(
+        _plugins.Remove(plugin);
+    }
+    
+    public void PluginSetData(PluginX plugin, string? data)
+    {
+        PluginData newData = new() {
+            PluginId = plugin.Id,
+            Data = data,
+        };
+        _pluginDataDb.Upsert(newData);
+    }
```
按字典序插入到别的函数后面吧，像之前接口说的那样

kurikomoe committed:
fix: add init for PluginX.Id

GoldenPotato137 commented on GalgameManager.WinApp.Base/Contracts/IPlugin.cs:
```diff
@@ -16,4 +18,18 @@ public interface IPlugin
+    /// </summary>
+    /// <remarks>
+    /// 对于开发阶段的插件可以在这个接口中实现热重载相关逻辑方便开发。
+    /// 插件默认最多用时 5s，可以通过 extendWaitHandler 延长，但是无法超过 60s
```
目前这个接口对于非dev插件不会被调用呀，最好还是修改一下注释😂

kurikomoe replied:
会调用，我忘了在 LoadPluginTask 里面调用一下 Uninstall 逻辑了。

kurikomoe replied:
另外感觉最好卸载的时候加一个全局模态对话框锁定一下 UI 会好一点，但是 UI 无能，咱把这个当心愿了

GoldenPotato137 commented on GalgameManager/Services/PluginService/PluginService.cs:
```diff
@@ -67,22 +192,26 @@ public async Task LoadPluginAsync(PluginX plugin, bool load)
 
         async void OnPluginOnPropertyChanged(object? sender, PropertyChangedEventArgs args)
         {
+            // FIXME(kuriko): 这里我们没有考虑 name 和 devMode 属性变化导致的插件列表重排序问题
```
把ob的排序函数挂上去应该就没这个问题了

kurikomoe committed:
fix: invoke plugin.uninstall in startup task
